.maintain/gitlab/check_companion_build.sh

```diff
@@ -0,0 +1,115 @@
+
+boldprint () { printf "|\n| \033[1m%s\033[0m\n|\n" "${@}"; }
+boldcat () { printf "|\n"; while read -r l; do printf "| \033[1m%s\033[0m\n" "${l}"; done; printf "|\n" ; }
+
+
+
+boldcat <<-EOT
+
+
+check_${REPO}_companion_build
+==============================
+
+this job checks if there is a string in the description of the pr like
+
+$REPO companion: $ORGANISATION/$REPO#567
+
+
+it will then run cargo check from this ${REPO}'s branch with substrate code
+from this pull request. otherwise, it will uses master instead
+
+
+EOT
+
+# Set the user name and email to make merging work
+git config --global user.name 'CI system'
+git config --global user.email '<>'
+
+# Merge master into our branch before building to make sure we don't miss
+# any commits that are required.
+git fetch --depth 100 origin
+git merge origin/master
+
+# Clone the current master branch into a local directory
+# NOTE: we need to pull enough commits to be able to find a common
+# ancestor for successfully performing merges below.
+git clone --depth 20 "https://github.com/${ORGANISATION}/${REPO}.git"
+
+cd "$REPO"
+
+# either it's a pull request then check for a companion otherwise use
+# master
+# shellcheck disable=SC2003
+if expr match "${CI_COMMIT_REF_NAME}" '^[0-9]\+$' >/dev/null
+then
+  boldprint "this is pull request no ${CI_COMMIT_REF_NAME}"
+
+  pr_data_file="$(mktemp)"
+  # get the last reference to a pr in the target repo
+  curl -sSL -H "${github_header}" -o "${pr_data_file}" \
+    "${github_api_substrate_pull_url}/${CI_COMMIT_REF_NAME}"
+
+  pr_body="$(sed -n -r 's/^[[:space:]]+"body": (".*")[^"]+$/\1/p' "${pr_data_file}")"
+
+  pr_companion="$(echo "${pr_body}" | sed -n -r \
+      -e "s;^.*[Cc]ompanion.*${ORGANISATION}/${REPO}#([0-9]+).*$;\1;p" \
+      -e "s;^.*[Cc]ompanion.*https://github.com/${ORGANISATION}/${REPO}/pull/([0-9]+).*$;\1;p" \
+    | tail -n 1)"
+
+  if [ "${pr_companion}" ]
+  then
+    boldprint "companion pr specified/detected: #${pr_companion}"
+    git fetch origin "refs/pull/${pr_companion}/head:pr/${pr_companion}"
+    git checkout "pr/${pr_companion}"
+    git merge origin/master
+  else
+    boldprint "no companion branch found - building ${REPO}:master"
+  fi
+  rm -f "${pr_data_file}"
+else
+  boldprint "this is not a pull request - building ${REPO}:master"
+fi
+
+# Patch polkadot, substrate or beefy deps as required
+declare -A match_arg
+match_arg=()
+match_arg["--substrate"]='source = "git+https://github.com/paritytech/substrate?'
+match_arg["--polkadot"]='source = "git+https://github.com/paritytech/polkadot?'
+match_arg["--beefy"]='source = "git+https://github.com/paritytech/parity-bridges-gadget?'
+
+# For each Cargo.lock
+while IFS= read -r cargo_lock; do
+  # If the Cargo.lock has a dependency, we patch with diener
+  for patch_arg in "${!match_arg[@]}"; do
+    if grep -q "${match_arg[$patch_arg]}" "$cargo_lock" ; then
+      echo "patching $patch_arg"
+      diener patch --crates-to-patch ../ "$patch_arg" --path Cargo.toml
```

Comment:
You might be executing diener multiple times for the same dependency if it's referred to in multiple different Cargo.lock files. Dunno if it's a problem but it would be cleaner to do like I had in the [original script](https://github.com/paritytech/substrate/pull/9010#discussion_r645511761). Instead of applying the patch directly inside of this loop, you'd just set the flag to true:

```bash
        patch_args["$patch_arg"]=true
```

Then, later, outside of this loop:

```bash
for patch_arg in "${!patch_args[@]}"; do
      echo "patching $patch_arg"
      diener patch --crates-to-patch ../ "$patch_arg" --path Cargo.toml
done
```

That way you only invoke Diener once per flag.